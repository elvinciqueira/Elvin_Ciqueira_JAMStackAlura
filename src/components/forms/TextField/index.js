import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import Typography from '../../foundation/Typography'

const InputWrapper = styled.div`
  margin-bottom: 17px;
`

const Input = styled(Typography)`
  width: 100%;
  border: 1px solid ${({theme}) => theme.colors.borders.main.color};
  padding: 8px 16px;
  outline: 0;
  margin-top: 8px;
  border-radius: 4px;
`

Input.defaultProps = {
  tag: 'input',
  variant: 'bodyText',
}

export default function TextField({
  placeholder,
  name,
  onChange,
  value,
  type,
  id,
  tag,
  label,
}) {
  return (
    <InputWrapper>
      <label htmlFor={name}>{label}</label>
      <Input
        tag={tag}
        id={id}
        type={type || 'text'}
        placeholder={placeholder}
        name={name}
        onChange={onChange}
        value={value}
      />
    </InputWrapper>
  )
}

TextField.propTypes = {
  placeholder: PropTypes.string,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
  label: PropTypes.string.isRequired,
  type: PropTypes.string,
  id: PropTypes.string.isRequired,
  tag: PropTypes.string,
}
