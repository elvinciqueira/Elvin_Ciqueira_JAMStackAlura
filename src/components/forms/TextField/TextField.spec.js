import React from 'react'
import user from '@testing-library/user-event'
import {render, screen} from '../../../infra/test/testUtils'
import TextField from './'

describe('<TextField />', () => {
  describe('when compononent renders', () => {
    test('should renders as an input', () => {
      render(
        <TextField
          id="name"
          tag="input"
          label="Seu nome"
          name="name"
          placeholder="Nome"
          type="text"
          value="Elvin"
          onChange={() => {}}
        />,
      )

      const textField = screen.getByPlaceholderText(/nome/i)

      expect(textField).toMatchSnapshot()
    })

    test('should renders as an textarea', () => {
      render(
        <TextField
          id="message"
          tag="textarea"
          label="Sua mensagem"
          name="message"
          placeholder="Mensagem"
          type="text"
          value="Elvin"
          onChange={() => {}}
        />,
      )

      const textField = screen.getByPlaceholderText(/mensagem/i)

      expect(textField).toMatchSnapshot()
    })

    test('should renders an input with type email', () => {
      render(
        <TextField
          id="email"
          tag="input"
          label="Seu E-mail"
          name="email"
          placeholder="Email"
          type="email"
          value="elvinciqueira@gmail.com"
          onChange={() => {}}
        />,
      )

      const textField = screen.getByPlaceholderText(/email/i)

      expect(textField).toMatchSnapshot()
    })

    test('should render an input with a label', () => {
      render(
        <TextField
          id="name"
          tag="input"
          label="Seu nome"
          name="name"
          placeholder="Nome"
          type="text"
          value="Elvin"
          onChange={() => {}}
        />,
      )

      const inputLabel = screen.getByLabelText(/seu nome/i)

      expect(inputLabel).toMatchSnapshot()
    })
  })

  describe('when field is valid', () => {
    describe('and user is typing', () => {
      test('the value must be updated', () => {
        const onChangeMock = jest.fn()

        render(
          <TextField
            id="name"
            tag="input"
            label="Seu nome"
            name="name"
            placeholder="Nome"
            type="text"
            value=""
            onChange={onChangeMock}
          />,
        )

        const inputNome = screen.getByPlaceholderText(/nome/i)
        user.type(inputNome, 'Elvin')

        expect(onChangeMock).toBeCalledTimes(5)
      })
    })
  })
})
