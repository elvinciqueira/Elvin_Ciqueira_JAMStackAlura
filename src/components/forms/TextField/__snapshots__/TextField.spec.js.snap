// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> when compononent renders should render an input with a label 1`] = `
.c0 {
  font-size: 1.125rem;
  font-weight: 400;
  font-family: Fira Sans,sans-serif;
}

.c1 {
  width: 100%;
  border: 1px solid #E9C46A;
  padding: 8px 16px;
  outline: 0;
  margin-top: 8px;
  border-radius: 4px;
}

<input
  class="c0 c1"
  id="name"
  name="name"
  placeholder="Nome"
  type="text"
  value="Elvin"
/>
`;

exports[`<TextField /> when compononent renders should renders an input with type email 1`] = `
.c0 {
  font-size: 1.125rem;
  font-weight: 400;
  font-family: Fira Sans,sans-serif;
}

.c1 {
  width: 100%;
  border: 1px solid #E9C46A;
  padding: 8px 16px;
  outline: 0;
  margin-top: 8px;
  border-radius: 4px;
}

<input
  class="c0 c1"
  id="email"
  name="email"
  placeholder="Email"
  type="email"
  value="elvinciqueira@gmail.com"
/>
`;

exports[`<TextField /> when compononent renders should renders as an input 1`] = `
.c0 {
  font-size: 1.125rem;
  font-weight: 400;
  font-family: Fira Sans,sans-serif;
}

.c1 {
  width: 100%;
  border: 1px solid #E9C46A;
  padding: 8px 16px;
  outline: 0;
  margin-top: 8px;
  border-radius: 4px;
}

<input
  class="c0 c1"
  id="name"
  name="name"
  placeholder="Nome"
  type="text"
  value="Elvin"
/>
`;

exports[`<TextField /> when compononent renders should renders as an textarea 1`] = `
.c0 {
  font-size: 1.125rem;
  font-weight: 400;
  font-family: Fira Sans,sans-serif;
}

.c1 {
  width: 100%;
  border: 1px solid #E9C46A;
  padding: 8px 16px;
  outline: 0;
  margin-top: 8px;
  border-radius: 4px;
}

<textarea
  class="c0 c1"
  id="message"
  name="message"
  placeholder="Mensagem"
  type="text"
>
  Elvin
</textarea>
`;
